### PR TITLE
with-redefs + wrap-with macros

### PR DESCRIPTION
**Pull request by caioaao**

### CODE REVIEW

File: src/state_flow/labs/state.clj [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+(defmacro with-redefs
+  [bindings flow]
+  (assert *enable-with-redefs-macro*
+          "`with-redefs` usage is not recommended. If you know what you're doing and really want to continue, set `*enable-with-redefs-macro*` to true")
+  `(m/do-let
+    [world#  (state/get)
+     runner# (state-flow/runner)
+     :let [[ret# state#] (clojure.core/with-redefs ~bindings
+                           (runner# ~flow world#))]]
+    (state/put state#)
```
Comment (dchelimsky) [inline src/state_flow/labs/state.clj:18]:
```suggestion
    (state-flow.api/swap-state (constantly state#))
```

File: src/state_flow/labs/state.clj [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    (state/return ret#)))
```
Comment (dchelimsky) [inline src/state_flow/labs/state.clj:19]:
```suggestion
    (state-flow.api/return ret#)))
```

**Comment by dchelimsky** [comment]
I think if we're going to add this to state-flow proper (even in labs), I think we should add the `wrap-in` generalization. The you could use e.g.

```clojure
(wrap-in (fn [wrapped-flow] (with-redefs [<bindings>] wrapped-flow))
         flow-to-wrap))
```

File: src/state_flow/labs/state.clj [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  (:refer-clojure :exclude [with-redefs])
+  (:require [cats.core :as m]
+            [state-flow.state :as state]
+            [state-flow.core :as state-flow]))
+
+(def ^:dynamic *enable-with-redefs-macro* nil)
+
+(defmacro with-redefs
+  [bindings flow]
+  (assert *enable-with-redefs-macro*
```
Comment (sovelten) [inline src/state_flow/labs/state.clj:11]: Not sure about the usage of this variable. Just a warning in the comments is fine. The flag looks like a handicap to me.

File: src/state_flow/labs/state.clj [hunk 4]
```diff
@@ -0,0 +1,19 @@
+(ns state-flow.labs.state
```
Comment (sovelten) [inline src/state_flow/labs/state.clj:1]: This would probably go better in another namespace. Maybe `labs.helpers`

**Comment by dchelimsky** [comment]
> I think if we're going to add this to state-flow proper (even in labs), I think we should add the wrap-in generalization.

Background: https://github.com/nubank/state-flow/pull/33

File: src/state_flow/labs/state.clj (hunk 4 above)
Comment (caioaao) [reply src/state_flow/labs/state.clj:1]: I thought it'd go along with the other state fns, like `wrap-fn`

**Comment by caioaao** [comment]
> I think if we're going to add this to state-flow proper (even in labs), I think we should add the wrap-in generalization. 

@dchelimsky from what I saw, #33 was closed because of the potential of misuse. If we limit this implementation to specific use cases we see the need for it, the misuse might be reduced, no? We can keep `wrap-with` but as a private function

**Comment by caioaao** [comment]
We can also leave it public for now in labs and check if this is going to be an issue

File: src/state_flow/labs/state.clj (hunk 4 above)
Comment (dchelimsky) [reply src/state_flow/labs/state.clj:1]: I think it's fine here, but we should have a docstring in this namespace that indicates that it is experimental and subject to change.

File: src/state_flow/labs/state.clj [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  (:refer-clojure :exclude [with-redefs])
+  (:require [cats.core :as m]
+            state-flow.api
+            [state-flow.core :as state-flow]
+            [state-flow.state :as state]))
+
+(defmacro wrap-with
+  "WARNING: `wrap-with` usage is not recommended. Use only if you know what you're
```
Comment (dchelimsky) [inline src/state_flow/labs/state.clj:9]: This warning is fine, but we should also say what it does :-D

File: src/state_flow/labs/state.clj [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  [wrapper-fn flow]
+  `(m/do-let
+    [world#  (state/get)
+     runner# (state-flow/runner)
+     :let [[ret# state#] (~wrapper-fn (fn [] (runner# ~flow world#)))]]
+    (state-flow.api/swap-state (constantly state#))
+    (state/return ret#)))
+
+(defmacro with-redefs
+  "WARNING: `with-redefs` usage is not recommended. Use only if you know what you're
```
Comment (dchelimsky) [inline src/state_flow/labs/state.clj:20]: Same comment - docstring should also say what this does, and warn that `with-redefs` is not threadsafe.

File: test/state_flow/labs/state_test.clj [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+(defn wrap-with-redefs [f]
+  (with-redefs [put2 put3]
+    (f)))
+
+(defn called-it-callback [f]
+  (print "called it")
+  (f))
+
+(deftest wrap-with-test
+  (testing "verify wrapper is called"
```
Comment (dchelimsky) [inline test/state_flow/labs/state_test.clj:20]:
```suggestion
  (testing "wrapper is called"
```

File: test/state_flow/labs/state_test.clj [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    (is (= "called it"
+           (with-out-str
+             (-> (labs.state/wrap-with
+                  (fn [f] (print "called it") (f))
+                  (state/modify put2))
+                 (state-flow/run {}))))))
+  (testing "verify flow runs successfully"
```
Comment (dchelimsky) [inline test/state_flow/labs/state_test.clj:27]:
```suggestion
  (testing "flow runs successfully"
```

File: test/state_flow/labs/state_test.clj [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    (is (match? [{} {:value 2}]
+                (-> (labs.state/wrap-with
+                  (fn [f] (f))
+                  (state/modify put2))
+                 (state-flow/run {}))))))
+
+(deftest with-redefs-test
+  (testing ""
```
Comment (dchelimsky) [inline test/state_flow/labs/state_test.clj:35]: Either fill in the docstring or bag `testing`

File: src/state_flow/labs/state.clj (hunk 5 above)
Comment (dchelimsky) [reply src/state_flow/labs/state.clj:9]: I think we should say that this here to support helper macros like `with-redefs`, and reference that as an implementation example.

**Comment by dchelimsky** [comment]
Thanks for all the updates @caioaao. I'm waiting for some feedback from other folks within Nubank.

**Comment by fernando-nubank** [comment]
> We can also leave it public for now in labs and check if this is going to be an issue

I have more issues with `with-redefs` than with `wrap-fn` but, as both of them are in `labs` I think they can be public.

**Comment by dchelimsky** [comment]
> LGTM now, but I want at least one other approval before merging.

Oh! There is one. I'm going to add a commit to update the changelog first. Then I'll merge it.

**Comment by dchelimsky** [comment]
> > LGTM now, but I want at least one other approval before merging.
> 
> Oh! There is one. I'm going to add a commit to update the changelog first. Then I'll merge it.

I can't do that on  your branch @caioaao so I'll do it in a follow up PR before releasing.

**Pull request merged**